neaten/neaten.go

```diff
@@ -0,0 +1,390 @@
+	}
+
+	return t.moveAndDelete()
+}
+
+// dirValid checks if the directory is valid: is absolute and exists.
+func dirValid(dir string) error {
+	dir, err := filepath.Abs(dir)
+	if err != nil {
+		return err
+	}
+
+	_, err = os.Stat(dir)
+
+	return err
+}
+
+// moveAndDelete does the main work of this package: it finds, renames and
+// moves the combine, base and db files, ensuring that their permissions
+// match those of our destDir.
+func (t *Tidy) moveAndDelete() error {
+	for inSuffix, outSuffix := range t.CombineFileSuffixes {
+		if err := t.findAndMoveOutputs(inSuffix, outSuffix); err != nil {
+			return err
+		}
+	}
+
+	for inSuffix, outSuffix := range t.DBFileSuffixes {
+		if err := t.findAndMoveDBs(inSuffix, outSuffix); err != nil {
+			return err
+		}
+	}
+
+	for inSuffix, outSuffix := range t.BaseFileSuffixes {
+		if err := t.moveBaseDirsFile(inSuffix, outSuffix); err != nil {
+			return err
+		}
+	}
+
+	return os.RemoveAll(t.SrcDir)
+}
+
+// findAndMoveOutputs finds output files in the given sourceDir with given
+// suffix and moves them to our destDir, including date in the name, and
+// adjusts ownership and permissions to match the destDir.
+func (t *Tidy) findAndMoveOutputs(inSuffix, outSuffix string) error {
+	outputPaths, err := filepath.Glob(fmt.Sprintf(t.CombineFileGlobPattern, t.SrcDir, inSuffix))
+	if err != nil {
+		return err
+	}
+
+	for _, path := range outputPaths {
+		err := t.moveOutput(path, outSuffix)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// moveOutput moves an output file to our desrDir and changes its name to
+// the correct format, then adjusts ownership and permissions to match the
+// destDir.
+func (t *Tidy) moveOutput(source string, suffix string) error {
+	interestUniqueDir := filepath.Dir(source)
+	interestBaseDir := filepath.Dir(interestUniqueDir)
+	multiUniqueDir := filepath.Dir(interestBaseDir)
+	dest := filepath.Join(t.DestDir, fmt.Sprintf("%s_%s.%s.%s.%s",
+		t.Date,
+		filepath.Base(interestBaseDir),
+		filepath.Base(interestUniqueDir),
+		filepath.Base(multiUniqueDir),
+		suffix))
+
+	return t.renameAndMatchPerms(source, dest)
+}
+
+// renameAndMatchPerms tries 2 ways to rename the file (resorting to a copy if
+// this is across filesystem boundaries), then matches the dest file permissions
+// to the given FileInfo.
+//
+// If source doesn't exist, but dest does, assumes the rename was done
+// previously and just tries to match the permissions.
+func (t *Tidy) renameAndMatchPerms(source, dest string) error {
+	if _, err := os.Stat(source); errors.Is(err, os.ErrNotExist) {
+		if _, err = os.Stat(dest); err == nil {
+			return t.PermsOK(dest)
+		}
+	}
+
+	err := os.Rename(source, dest)
+	if err != nil {
+		if err = shutil.CopyFile(source, dest, false); err != nil {
+			return err
+		}
+	}
+
+	return t.PermsOK(dest)
+}
+
+// PermsOK checks whether the given file has the same ownership and read-write
+// permissions as our destDir.
+func (t *Tidy) PermsOK(path string) error {
```

Comment:
Name and description of this function is wrong. It doesn't just check the perms are the same; if they're not it changes the permissions to match. Try CorrectPerms().

Likewise for the sub functions it calls for owners and readWrite